drivers/input/input_adc_keys.c

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (c) 2024 Chen Xingyu <hi@xingrz.me>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT adc_keys
+
+#include <stdlib.h>
+#include <stdbool.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/adc.h>
+#include <zephyr/input/input.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/util.h>
+
+LOG_MODULE_REGISTER(adc_keys, CONFIG_INPUT_LOG_LEVEL);
+
+struct adc_keys_code_config {
+	uint32_t key_index;
+	int32_t press_mv;
+};
+
+struct adc_keys_key_state {
+	bool last_state;
+	bool curr_state;
+};
+
+struct adc_keys_config {
+	struct adc_dt_spec channel;
+	uint32_t sample_period_ms;
+	int32_t keyup_mv;
+	const struct adc_keys_code_config *code_cfg;
+	uint32_t code_cnt;
+	const uint16_t *key_code;
+	struct adc_keys_key_state *key_state;
+	uint32_t key_cnt;
+};
+
+struct adc_keys_data {
+	const struct device *self;
+	struct k_work_delayable dwork;
+	struct adc_sequence seq;
+};
+
+static inline int32_t adc_keys_read(const struct device *dev)
```

Comment:
`uint32_t` for return value?

Reply:
The value given by `adc_raw_to_millivolts_dt` is of type `int32_t`. Is it worth casting it to `uint32_t`? Anyway, I think it's no difference in assembly level.

Reply:
ok fair enough, nah let's keep this signed